### PR TITLE
Convert Homebrew::Hooks::Bottles test to spec.

### PR DESCRIPTION
**Pull request by reitermarkus**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [x] Have you successfully run `brew tests` with your changes locally?

-----

Convert tests to RSpec.

### CODE REVIEW

File: Library/Homebrew/test/bottle_hooks_spec.rb [hunk 1]
```diff
@@ -0,0 +1,51 @@
+require "formula_installer"
+require "hooks/bottles"
+
+describe Homebrew::Hooks::Bottles do
+  subject { formula_installer }
+
+  let(:formula_installer) {
```
Comment (MikeMcQuaid) [inline Library/Homebrew/test/bottle_hooks_spec.rb]: How about doing `let(:formula_installer) { FormulaInstaller.new formula }` for one-liners < 80 characters?

File: Library/Homebrew/test/bottle_hooks_spec.rb [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    FormulaInstaller.new formula
+  }
+
+  let(:formula) {
+    double(bottle_disabled?: false, some_random_method: true, bottle: nil, local_bottle_path: nil)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/test/bottle_hooks_spec.rb]: Probably worth splitting out the hash.

File: Library/Homebrew/test/bottle_hooks_spec.rb [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+  end
+
+  describe "has bottle" do
+    let(:block) { :some_random_method }
+    it { is_expected.to be_pour_bottle }
+  end
+
+  describe "has no bottle" do
+    let(:block) {
+      ->(f) { !f.some_random_method }
```
Comment (MikeMcQuaid) [inline Library/Homebrew/test/bottle_hooks_spec.rb]: I have no idea what this is doing 😁. Can you explain?

File: Library/Homebrew/test/build_environment_spec.rb [hunk 4]
```diff
@@ -0,0 +1,70 @@
+require "build_environment"
+
+RSpec::Matchers.define :use_userpaths do
+  match do |actual|
```
Comment (MikeMcQuaid) [inline Library/Homebrew/test/build_environment_spec.rb]: `match {|actual| actual.userpaths?`?

File: Library/Homebrew/test/bottle_hooks_spec.rb (hunk 1 above)
Comment (MikeMcQuaid) [inline Library/Homebrew/test/bottle_hooks_spec.rb]: Also, why not `let do`?

File: Library/Homebrew/test/build_environment_spec.rb [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+  subject {
+    Object.new.extend(BuildEnvironmentDSL).tap { |o| o.instance_eval(&block) }
+  }
+
+  context "single argument" do
+    let(:block) {
+      proc do
+        env :userpaths
+      end
+    }
```
Comment (alyssais) [inline Library/Homebrew/test/build_environment_spec.rb]: This proc-inside-a-block thing is weird, especially with the `do`/`end` inside the `{`/`}`.

Maybe just define it as a method?:

```ruby
def block
  env :userpaths
end
```

and evaluate it with

```ruby
o.instance_eval(&method(:block))
```

File: Library/Homebrew/test/build_environment_spec.rb [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+    it "returns false if the environment does not contain :userpaths" do
+      expect(env).not_to use_userpaths
+    end
+  end
+end
+
+describe BuildEnvironmentDSL do
+  subject {
+    Object.new.extend(BuildEnvironmentDSL).tap { |o| o.instance_eval(&block) }
```
Comment (alyssais) [inline Library/Homebrew/test/build_environment_spec.rb]: I think this should be split up into multiple lines.

The subject line should just construct the subject object, and the block should either be evaluated inside a `before` or inside the tests themselves.

File: Library/Homebrew/test/build_environment_spec.rb [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    actual.userpaths?
+  end
+end
+
+describe BuildEnvironment do
+  let(:env) { described_class.new }
+
+  describe "#<<" do
+    it "returns itself" do
+      expect(env << :foo).to eq env
```
Comment (alyssais) [inline Library/Homebrew/test/build_environment_spec.rb]: You've changed the behaviour of these tests. The Minitest equivalents used `assert_same`, which checks instance equality using `#equal?`. The equivalent construction in RSpec is `be` (instead of `eq`).

File: Library/Homebrew/test/build_environment_spec.rb [hunk 8]
```diff
@@ -0,0 +1,70 @@
+require "build_environment"
+
+RSpec::Matchers.define :use_userpaths do
+  match do |actual|
+    actual.userpaths?
+  end
+end
+
+describe BuildEnvironment do
+  let(:env) { described_class.new }
```
Comment (alyssais) [inline Library/Homebrew/test/build_environment_spec.rb]: Should this not be `subject`?

File: Library/Homebrew/test/build_environment_spec.rb [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+  describe "#merge" do
+    it "returns itself" do
+      expect(env.merge([])).to eq env
+    end
+  end
+
+  describe "#std?" do
+    it "returns true if the environment contains :std" do
+      expect(env << :std).to be_std
```
Comment (alyssais) [inline Library/Homebrew/test/build_environment_spec.rb]: Could this and similar tests be split up into multiple lines? I don't think mutation/setup should happen _inside_ an expectation.

These are specs — we want them to be as simple as possible to read and understand.

File: Library/Homebrew/test/build_environment_spec.rb [hunk 10]
```diff
@@ -0,0 +1,70 @@
+require "build_environment"
+
+RSpec::Matchers.define :use_userpaths do
```
Comment (alyssais) [inline Library/Homebrew/test/build_environment_spec.rb]: There should actually be no need to define a new matcher here at all:

```ruby
RSpec::Matchers.alias_matcher :use_userpaths, :be_userpaths
```

File: Library/Homebrew/.rubocop.yml [hunk 11]
```diff
@@ -11,8 +11,8 @@ AllCops:
 
 Style/BlockDelimiters:
   Exclude:
-    - '**/cask/spec/**/*'
-    - '**/cask/test/**/*'
+    - '**/spec/**/*'
+    - '**/test/**/*'
```
Comment (alyssais) [inline Library/Homebrew/.rubocop.yml]: Can you explain what this does and why it was needed?

File: Library/Homebrew/test/build_environment_spec.rb (hunk 8 above)
Comment (reitermarkus) [inline Library/Homebrew/test/build_environment_spec.rb]: I think in this it's nicer to have a named variable since there is no good way to use the short-hand `it { is_expected }` syntax.

File: Library/Homebrew/test/build_environment_spec.rb (hunk 10 above)
Comment (reitermarkus) [inline Library/Homebrew/test/build_environment_spec.rb]: I searched for something like this but couldn't find it, thanks.

File: Library/Homebrew/test/build_environment_spec.rb (hunk 5 above)
Comment (reitermarkus) [inline Library/Homebrew/test/build_environment_spec.rb]: I don't like defining methods inside specs, that is specifically what procs and lambda are for. Anyway, in this case it really only works with procs, not with lambdas or methods.

File: Library/Homebrew/test/build_environment_spec.rb (hunk 8 above)
Comment (alyssais) [inline Library/Homebrew/test/build_environment_spec.rb]: That's true, but there's also some semantic meaning attached to `subject` and I think consistency is desirable here — it's not just a hack you use to get nice syntax. When I glanced at these tests I assumed `env` to be some kind of helper object.

**Comment by alyssais** [comment]
Additionally, could the conversion be done across multiple PRs, maybe one per file? If the whole test suite is rewritten it's going to be really hard to keep up with and review effectively (and it'll take forever to get any of it merged!)

File: Library/Homebrew/test/bottle_hooks_spec.rb (hunk 3 above)
Comment (reitermarkus) [inline Library/Homebrew/test/bottle_hooks_spec.rb]: It's passing the lambda to `setup_formula_has_bottle`.

File: Library/Homebrew/.rubocop.yml (hunk 11 above)
Comment (MikeMcQuaid) [inline Library/Homebrew/.rubocop.yml]: Agreed.

File: Library/Homebrew/test/build_environment_spec.rb (hunk 5 above)
Comment (MikeMcQuaid) [inline Library/Homebrew/test/build_environment_spec.rb]: I agree. It may be what procs and lambda are for but it makes the tests significantly more opaque for beginning or non-Rubyists.

File: Library/Homebrew/test/bottle_hooks_spec.rb [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+require "hooks/bottles"
+
+RSpec::Matchers.alias_matcher :pour_bottle, :be_pour_bottle
+
+describe Homebrew::Hooks::Bottles do
+  subject { formula_installer }
+
+  let(:formula_installer) { FormulaInstaller.new formula }
+
+  let(:formula) {
```
Comment (MikeMcQuaid) [inline Library/Homebrew/test/bottle_hooks_spec.rb]: Can we use `let do` on multiline lets?

File: Library/Homebrew/test/bottle_hooks_spec.rb [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+    double(
+      bottle: nil,
+      local_bottle_path: nil,
+      bottle_disabled?: false,
+      some_random_method: true
```
Comment (MikeMcQuaid) [inline Library/Homebrew/test/bottle_hooks_spec.rb]: May as well add a trailing comma here.

**Comment by MikeMcQuaid** [comment]
I think the heavy use of procs, lambdas and subject makes this significantly less readable than it would be using methods and duplicating function calls.

File: Library/Homebrew/.rubocop.yml (hunk 11 above)
Comment (reitermarkus) [inline Library/Homebrew/.rubocop.yml]: This is so RuboCop doesn't complain about multi-line `expect`s, e.g.:

```ruby
expect {
  # something
}.not_to raise_error
```

File: Library/Homebrew/test/bottle_hooks_spec.rb (hunk 13 above)
Comment (reitermarkus) [inline Library/Homebrew/test/bottle_hooks_spec.rb]: RuboCop complains about commas after the last parameter of a method call.

Comment (MikeMcQuaid) [inline Library/Homebrew/test/bottle_hooks_spec.rb]: Boo. Ignore me.

File: Library/Homebrew/.rubocop.yml (hunk 11 above)
Comment (MikeMcQuaid) [inline Library/Homebrew/.rubocop.yml]: Boo. Ignore me.

**Comment by MikeMcQuaid** [comment]
> I think the heavy use of procs, lambdas and subject makes this significantly less readable than it would be using methods and duplicating function calls.

And to clarify on this: I think making it less readable for newer Rubyists will make us less likely to have good, new tests implemented by new contributors.

**Comment by reitermarkus** [comment]
Removed the `proc`s and put `instance_eval` inside `before` blocks instead.

File: Library/Homebrew/test/bottle_hooks_spec.rb (hunk 13 above)
Comment (alyssais) [inline Library/Homebrew/test/bottle_hooks_spec.rb]: Perhaps that's something we should re-visit? Trailing commas are very handy when argument lists are on multiple lines IMO.

Comment (MikeMcQuaid) [inline Library/Homebrew/test/bottle_hooks_spec.rb]: If it's a rule that can be tweaked 👍

File: Library/Homebrew/test/bottle_hooks_spec.rb [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+
+    context "given a block which evaluates to false" do
+      let(:block) { ->(f) { !f.some_random_method } }
+
+      it { is_expected.not_to pour_bottle }
+    end
+  end
+
+  describe "#setup_pour_formula_bottle" do
+    subject { -> { formula_installer.pour } }
```
Comment (MikeMcQuaid) [inline Library/Homebrew/test/bottle_hooks_spec.rb]: What would this look like without a proc/lambda?

File: Library/Homebrew/test/bottle_hooks_spec.rb [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+
+    let(:block) { ->(_) { true } }
```
Comment (MikeMcQuaid) [inline Library/Homebrew/test/bottle_hooks_spec.rb]: What would this look like without a proc/lambda?

File: Library/Homebrew/test/build_environment_spec.rb [hunk 16]
```diff
@@ -0,0 +1,66 @@
+require "build_environment"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/test/build_environment_spec.rb]: I'm 👍 on this test file being merged if you split into a separate PR.

**Comment by alyssais** [comment]
Very happy with how this turned out 🙂
